src/systems/filecoin_blockchain/storage_power_consensus/_index.md

```diff
@@ -44,70 +46,213 @@ More precisely,
+         // block validation logic.
+         returns getDrandEntryFromBlockHeader(chain.Block(minEpoch))
+    // otherwise, we need to continue progressing on the chain, i.e. maybe no
+    // miner were elected or filecoin / drand outage
+    minEpoch++
+  }
+}
+
+func getDrandEntryFromBlockHeader(block,round) (DrandEntry,error) {
+    for _,dr := range block.DrandEntries {
+        if dr.Round == round {
+            return dr
+        }
+    }
+    return errors.New("drand entry not found in block")
+}
+```
 
-In Filecoin, every block header contains a single ticket.
+### Fetch randomness from drand network
 
-You can find the Ticket data structure {{<sref data_structures "here">}}.
+When mining, a miner can fetch entries from the drand network to include them in
+the new block by calling the method `GetBeaconEntriesForEpoch`.
+```
+GetBeaconEntriesForEpoch(epoch) []BeaconEntry {
+
+    // special case genesis: the genesis block is pre-generated and so cannot include a beacon entry 
+    // (since it will not have been generated). Hence, we only start checking beacon entries at the first block after genesis.
+    // If that block includes a wrong beacon entry, we simply assume that a majority of honest miners at network birth will
+    // simply fork.
+    entries := []
+    if epoch == 0 {
+        return entries
+    }
+
+    maxDrandRound := MaxBeaconRoundForEpoch(epoch)
+
+    // if checking the first post-genesis block, simply fetch the latest entry.
+    if epoch == 1 {
+        rand := drand.Public(maxDrandRound)
+        return append(entries, rand)
+    }
+
+    // for the rest, fetch all drand entries generated between this epoch and last
+    prevMaxDrandRound := MaxBeaconRoundForEpoch(epoch - 1)
+    if (maxDrandRound == prevMaxDrandRound) {
+        // no new beacon randomness
+        return entries
+    }
+
+    entries := []
+    curr := maxDrandRound
+    for curr > prevMaxDrandRound {
+        rand := drand.Public(curr)
+        entries = append(entries, rand)
+        curr -= 1
+    }
+    // return entries in increasing order
+    reverse(entries)
+    return entries
+}
+```
 
-### Comparing Tickets in a Tipset and using ticket values
 
-Whenever comparing tickets is evoked in Filecoin, for instance when discussing selecting the "min ticket" in a Tipset, the comparison is that of the ticket's VRFDigest's bytes.
+{{<label beacon_entry_validation>}}
+### Validating Beacon Entries on block reception
+
+Per the above, a Filecoin chain will contain the entirety of the beacon's output from the Filecoin genesis to the current block.
+
+Given their role in leader election and other critical protocols in Filecoin, a block's beacon entries must be validated for every block. See {{<sref drand>}} for details. This can be done by ensuring every beacon entry is a valid signature over the prior one in the chain, using drand's [`Verify`](https://github.com/drand/drand/blob/master/beacon/beacon.go#L72) endpoint as follows:
```

Comment:
The source file in drand with the Verify function has moved and the link 404s. I'm gonna updated it to point to https://github.com/drand/drand/blob/763e9a252cf59060c675ced0562e8eba506971c1/chain/beacon.go#L76